x-pack/plugins/actions/server/lib/get_token_count_from_invoke_stream.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { encode } from 'gpt-tokenizer';
+import { Readable } from 'stream';
+import { finished } from 'stream/promises';
+
+export interface InvokeBody {
+  messages: Array<{
+    role: string;
+    content: string;
+  }>;
+}
+
+/**
+ * Takes the OpenAI and Bedrock `invokeStream` sub action response stream and the request messages array as inputs.
+ * Uses gpt-tokenizer encoding to calculate the number of tokens in the prompt and completion parts of the response stream
+ * Returns an object containing the total, prompt, and completion token counts.
+ * @param responseStream the response stream from the `invokeStream` sub action
+ * @param body the request messages array
+ */
+export async function getTokenCountFromInvokeStream({
+  responseStream,
+  body,
+}: {
+  responseStream: Readable;
+  body: InvokeBody;
+}): Promise<{
+  total: number;
+  prompt: number;
+  completion: number;
+}> {
+  const chatCompletionRequest = body;
+
+  // per https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+  const promptTokens = encode(
+    chatCompletionRequest.messages
+      .map((msg) => `<|start|>${msg.role}\n${msg.content}<|end|>`)
+      .join('\n')
+  ).length;
+
+  let responseBody: string = '';
+
+  responseStream.on('data', (chunk: string) => {
+    responseBody += chunk.toString();
+  });
+
+  try {
+    await finished(responseStream);
+  } catch {
+    // no need to handle this explicitly
```

Comment:
If we silent the error, the try/catch in the call [here](https://github.com/elastic/kibana/pull/170815/files#diff-7678f7e9330ad27120896f6a32fff43c1ad338f8a02eb2558f4f77ee320f7d65R43) won't catch the error either. How will we notice if there is an error here?

Reply:
True, I was copying from the openai implementation. I will update it there too